File: python/scenario/_tracing/sdk_metadata.py
"""SDK identity constants for scenario trace attributes.

Mirrors the TypeScript implementation in
``javascript/src/tracing/sdk-metadata.ts`` (merged PR #736).

These constants are stamped on every ``Scenario Turn`` root span so a trace
can be identified as produced by ``langwatch-scenario`` without asking the
user to re-derive it.
"""

from importlib.metadata import version, PackageNotFoundError

# ---------------------------------------------------------------------------
# Attribute key constants (shared so callers never type string literals)
# ---------------------------------------------------------------------------

ATTR_SCENARIO_SDK_NAME = "scenario.sdk.name"
"""OpenTelemetry attribute key for the scenario SDK name."""

ATTR_SCENARIO_SDK_VERSION = "scenario.sdk.version"
"""OpenTelemetry attribute key for the scenario SDK version."""

# ---------------------------------------------------------------------------
# Values
# ---------------------------------------------------------------------------

SCENARIO_SDK_NAME: str = "langwatch-scenario"
"""The distribution name of the Python scenario SDK."""


def sdk_version() -> str:
    """Return the installed ``langwatch-scenario`` package version.

    Reads from ``importlib.metadata`` at call time so it reflects the
    actual installed version rather than a hardcoded literal.

    Falls back to ``"unknown"`` when the package metadata is not available
    (e.g. editable installs without a dist-info directory) so a missing
    version never crashes a run.
    """
    try:
        return version(SCENARIO_SDK_NAME)
    except PackageNotFoundError:
        return "unknown"


#: Pre-computed version string — resolved once at module import.
SCENARIO_SDK_VERSION: str = sdk_version()

#: Convenience dict of both SDK-identity attributes, computed once.
scenario_sdk_attributes: dict[str, str] = {
    ATTR_SCENARIO_SDK_NAME: SCENARIO_SDK_NAME,
    ATTR_SCENARIO_SDK_VERSION: SCENARIO_SDK_VERSION,
}

File: javascript/src/tracing/sdk-metadata.ts
// Read the package version from package.json so it stays a single source of
// truth. tsup/esbuild inlines this JSON import at build time into a literal,
// mirroring how the langwatch observability SDK reads its own `version`
// (`import { version } from "../../package.json"`). This keeps ESM and CJS
// builds working with no custom build step and no `fs`/`require` at runtime.
import pkg from "../../package.json";

/**
 * OpenTelemetry attribute key for the scenario SDK name.
 * Mirrors langwatch's `langwatch.sdk.name` naming convention.
 */
export const ATTR_SCENARIO_SDK_NAME = "scenario.sdk.name";

/**
 * OpenTelemetry attribute key for the scenario SDK version.
 * Mirrors langwatch's `langwatch.sdk.version` naming convention.
 */
export const ATTR_SCENARIO_SDK_VERSION = "scenario.sdk.version";

/**
 * The scenario SDK name, stamped on every scenario-emitted trace so a run can
 * be identified as produced by `@langwatch/scenario` from its trace alone.
 * Sourced from package.json (single source of truth), not a hardcoded literal.
 */
export const SCENARIO_SDK_NAME: string = pkg.name;

/**
 * The installed `@langwatch/scenario` version, read from package.json.
 *
 * Note: no `scenario.sdk.commit` is emitted. A commit SHA is not cleanly
 * available at build/publish time (package.json carries no `gitHead`, and there
 * is no git context during publish without adding a custom build step), so it
 * is intentionally omitted. See issue #733 investigation.
 */
export const SCENARIO_SDK_VERSION: string = pkg.version;

/**
 * The SDK-identity attributes stamped on scenario trace spans (e.g. the
 * top-level `Scenario Turn` span). Attaching these as span attributes — rather
 * than resource attributes — means the identity travels with the span whenever
 * it is exported, covering both the SDK-owned exporter path and the case where
 * a consuming app configures its own OTEL exporter.
 *
 * Computed once at module load (the values are constants), so referencing it
 * costs no per-span allocation.
 */
export const scenarioSdkAttributes: Readonly<Record<string, string>> = {
  [ATTR_SCENARIO_SDK_NAME]: SCENARIO_SDK_NAME,
  [ATTR_SCENARIO_SDK_VERSION]: SCENARIO_SDK_VERSION,
};
